### PR TITLE
Golang update, expose Ceph versions and features, PG inactive tracking

### PR DESCRIPTION
**Pull request by Matt1360**
A few niceties:

- Update the golang version to 1.16.3
- Fix struct tag duplication in `OSDLabel`
- Expose `ceph versions` (generic) and `ceph features` (Nautilus) as metrics (tests pending)
- Track how long `ceph_exporter` has been aware of the oldest lingering PG that is not `active`, which aids in differentiating between a rolling peering storm and a real problem.

### CODE REVIEW

File: collectors/monitors.go [hunk 1]
```diff
@@ -371,6 +448,29 @@ func (m *MonitorCollector) collect() error {
...
+	// In a consistent cluster, there will only be one iteration (and label set) per daemon.
+	for daemon, vers := range versions {
+		for version, num := range vers {
+			// We have a version, which is something like the following, how we want to map it
+			// ceph version 12.2.13-30-aabbccdd (c5b1fd521188ddcdedcf6f98ae0e6a02286042f2) luminous (stable)
+			//              version_tag          sha1                                      release_tag
+			res := versionRegexp.FindStringSubmatch(version)
+			if len(res) != 4 {
+				continue
+			}
```
Comment (baergj) [inline collectors/monitors.go:461]: I wonder if we want to report these with labels set to `unknown` or something like that?

File: collectors/monitors.go [hunk 2]
```diff
@@ -371,6 +448,29 @@ func (m *MonitorCollector) collect() error {
...
+
+			m.CephVersions.WithLabelValues(daemon, res[1], res[2], res[3]).Set(float64(num))
```
Comment (baergj) [inline collectors/monitors.go:463]: For clarity you could use https://golang.org/pkg/regexp/#Regexp.SubexpIndex here, but it also makes this much wordier.

File: collectors/osd.go [hunk 3]
```diff
@@ -127,6 +131,11 @@ type OSDCollector struct {
 
 	// OSDObjectsBackfilled displays average number of objects backfilled in an OSD
 	OSDObjectsBackfilled *prometheus.CounterVec
+
+	// OldestInactivePG gives us the amount of time that the oldest inactive PG
+	// has been inactive for.  This is useful in rolling peering storms, or if
+	// you're cancelling a lot of backfill.  See `pgremapper` for more :)
```
Comment (baergj) [inline collectors/osd.go:137]: I would reword the `This is useful` sentence to something like: `This is useful to discern between rolling peering (such as when issuing a bunch of upmaps or weight changes) and a single PG stuck peering, for example.`

File: collectors/osd.go [hunk 4]
```diff
@@ -1107,6 +1127,41 @@ func (o *OSDCollector) cephPGQueryCommand(pgid string) []byte {
 	return cmd
 }
 
+func (o *OSDCollector) collectPGStates(ch chan<- prometheus.Metric) error {
+	// - See if there are OSDs in our tracking that are now active
```
Comment (baergj) [inline collectors/osd.go:1131]: This says OSDs but it's PGs.

File: collectors/osd.go [hunk 5]
```diff
@@ -1107,6 +1127,41 @@ func (o *OSDCollector) cephPGQueryCommand(pgid string) []byte {
...
+		active := strings.Contains(pg.State, "active")
+		_, ok := o.oldestInactivePGMap[pg.PGID]
+
+		// If we were tracking it, and it's now active, remove it
+		if active {
+			if ok {
+				delete(o.oldestInactivePGMap, pg.PGID)
+			}
+			continue
+		}
```
Comment (baergj) [inline collectors/osd.go:1147]: Minor simplification - I would suggest:
```go
active := strings.Contains(pg.State, "active")
if active {
    delete(o.oldestInactivePGMap, pg.PGID)
    continue
}

pgTime, ok := o.oldestInactivePGMap[pg.PGID]
if !ok {
    pgTime = now
    o.oldestInactivePGMap[pg.PGID] = now
}

if pgTime.Before(oldestTime) {
    oldestTime = pgTime
}
```

**Review by baergj — COMMENTED** [review]
Just some minor comments. There should probably be unit tests for the new stats.

File: collectors/monitors.go [hunk 6]
```diff
@@ -371,6 +503,30 @@ func (m *MonitorCollector) collect() error {
...
+	// Ceph versions, one loop for each daemon.
+	// In a consistent cluster, there will only be one iteration (and label set) per daemon.
+	for daemon, vers := range versions {
+		for version, num := range vers {
+			// We have a version, which is something like the following, how we want to map it
+			// ceph version 12.2.13-30-aabbccdd (c5b1fd521188ddcdedcf6f98ae0e6a02286042f2) luminous (stable)
+			//              version_tag          sha1                                      release_tag
+			res := versionRegexp.FindStringSubmatch(version)
+			if len(res) != 4 {
+				m.CephVersions.WithLabelValues(daemon, "unknown", "unknown", "unknown").Set(num)
```
Comment (baergj) [inline collectors/monitors.go:515]: If we have multiple of these within the daemon, they'll stomp on each other. Not sure if it really matters enough to account for that, though.

**Review by baergj — APPROVED** [review]

File: collectors/monitors.go (hunk 6 above)
Comment (Matt1360) [reply collectors/monitors.go:515]: Yeah, if it's unknown, it might not be a release that we should be trying to support (older than Luminous, or newer than anything what's available with today).
